### ISSUE TITLE
"begin...rescue vs respond_to? for Control Flow" inaccurate if method exists

### ISSUE THREAD

**Issue by jrunning** [JuanitoFatas/fast-ruby#81]
The section ["`begin...rescue` vs `respond_to?` for Control Flow"](https://github.com/JuanitoFatas/fast-ruby#beginrescue-vs-respond_to-for-control-flow-code) is correct when the receiver does not respond to the method (because rescuing an exception is expensive), but when the receiver *does* respond to the method the `begin...rescue` idiom is much faster—in my tests about 15x faster.

You can see my tests and their results here: https://gist.github.com/jrunning/0e18236c6363af15c520 I tested a number of different scenarios: implicit receiver (LocalVariableOrMethod), explicit receiver, a long inheritance chain and implicit receiver with "global" methods, but in all cases it was pretty much the same:

```text
Comparison:
test_rescue_global_does_exist:  7012056.7 i/s
test_respond_to_global_does_exist:  5073566.7 i/s - 1.38x slower
test_respond_to_global_doesnt_exist:  5019149.8 i/s - 1.40x slower
test_rescue_global_doesnt_exist:   464031.9 i/s - 15.11x slower
```

In other words, `rescue` is fastest when the method does exist, `respond_to?` is 1.38–1.48x slower whether the method exists or not, and `rescue` is 15–17x slower when the method doesn't exist.

With this in mind, I think it's harder to give definitive advice on  `rescue` vs. `respond_to?`. If one is looking for the fastest idiom then it depends on whether they think the method is more likely to exist or to be missing.

**Comment by nirvdrum** [comment]
I realize this project has a definite MRI slant to it, but I'd be remiss if I didn't mention that JRuby+Truffle can essentially compile away the `respond_to?` call by caching the interior method lookup. It's basically multi-dimensional inline cache and at that point, you get all the speculative optimizations that an inline cache affords. I [presented the work](https://www.youtube.com/watch?v=lRMWwjqbXUo) at the 2015 RubyKaigi with the intention of getting MRI to adopt a similar mechanism for fast metaprogramming.

Apologies if this looks like I'm highjacking the thread. I guess I'm just suggesting there's yet another variable that influences the results on alternative implementations and may influence MRI in a future release.

**Comment by jrunning** [comment]
@nirvdrum I think that's definitely relevant, though I don't know if this project's maintainers would agree. If it's possible to make Travis run JRuby+Truffle I'd definitely +1 a PR to add it to the [`.travis.yml`](https://github.com/JuanitoFatas/fast-ruby/blob/master/.travis.yml).

**Comment by mateusdeap** [comment]
For now we're trying to tidy up the project, but we can definitely look into providing benchmarks using other ruby implementations.

Closing this since the PR has been merged.
